File: backend/app/services/aluno_service.rb
# frozen_string_literal: true

class AlunoService
  def initialize(firestore_service: FirestoreService.new, file_upload_service: FileUploadService.new)
    @firestore = firestore_service
    @file_upload = file_upload_service
  end

  def buscar_perfil(uid, email, usuario_doc: nil)
    usuario = resolve_usuario(uid, email, usuario_doc)
    vinculo = primeiro_vinculo(usuario)

    {
      uid: FirestoreKeyNormalizer.fetch_field(usuario[:data], "uid").presence || usuario[:id],
      nome: FirestoreKeyNormalizer.fetch_field(usuario[:data], "nome"),
      email: FirestoreKeyNormalizer.fetch_field(usuario[:data], "email"),
      telefone: "",
      ingresso: "",
      curso: curso_do_vinculo(vinculo),
      departamento: "",
      matricula: texto(vinculo["matricula"])
    }
  end

  def buscar_resumo(uid, email, usuario_doc: nil)
    usuario = resolve_usuario(uid, email, usuario_doc)
    vinculo = primeiro_vinculo(usuario)
    atividades = listar_historico(uid, email, usuario_doc: usuario)

    aprovadas = atividades.count { |a| a[:status] == "APROVADO" }
    pendentes = atividades.count { |a| a[:status] == "PENDENTE" }
    indeferidas = atividades.count { |a| %w[REPROVADO INDEFERIDO].include?(a[:status]) }

    saldos = FirestoreKeyNormalizer.fetch_field(vinculo, "saldos") || {}
    horas_concluidas = FirestoreKeyNormalizer.fetch_numeric(saldos, "ensino") +
                       FirestoreKeyNormalizer.fetch_numeric(saldos, "pesquisa") +
                       FirestoreKeyNormalizer.fetch_numeric(saldos, "extensao")
    horas_necessarias = FirestoreKeyNormalizer.fetch_numeric(vinculo, "ch_total_exigida")
    horas_necessarias = 200 if horas_necessarias <= 0

    percentual = horas_necessarias.positive? ? [100, (horas_concluidas * 100) / horas_necessarias].min : 0

    {
      curso: curso_do_vinculo(vinculo),
      horasConcluidas: horas_concluidas,
      horasNecessarias: horas_necessarias,
      percentualConcluido: percentual,
      aprovadas: aprovadas,
      pendentes: pendentes,
      indeferidas: indeferidas,
      totalAtividades: atividades.size
    }
  end

  def listar_historico(uid, email, usuario_doc: nil)
    usuario = resolve_usuario(uid, email, usuario_doc)
    docs = @firestore.listar_solicitacoes_por_aluno(uid, usuario_doc: usuario)
    docs.map { |doc| para_atividade(doc) }.sort_by { |a| a[:data].to_s }.reverse
  end

  def listar_recentes(limite, uid, email, usuario_doc: nil)
    listar_historico(uid, email, usuario_doc: usuario_doc).first(limite)
  end

  def submeter_atividade(payload, uid, email)
    validar_submissao(payload)
    usuario = resolve_usuario(uid, email)
    vinculo = primeiro_vinculo(usuario)
    id_curso = payload["idCurso"].to_s.strip.presence ||
               texto(FirestoreKeyNormalizer.fetch_field(vinculo, "id_curso"))
    categoria = resolver_categoria(payload["categoria"], payload["tipo"])
    comprovante_url = @file_upload.upload_data_url(payload["comprovanteUrl"], usuario[:id])

    doc = {
      "uid_aluno" => FirestoreKeyNormalizer.fetch_field(usuario[:data], "uid").presence || usuario[:id],
      "id_curso" => FirestoreKeyNormalizer.normalize(id_curso).presence || id_curso,
      "titulo_atividade" => payload["titulo"].to_s.strip,
      "categoria" => categoria,
      "horas_informadas" => payload["horas"].to_i,
      "horas_aprovadas" => 0,
      "status" => "pendente",
      "url_certificado" => comprovante_url,
      "data_envio" => Time.current,
      "justificativa_coordenador" => ""
    }

    salvo = @firestore.salvar_solicitacao(doc)
    para_atividade(salvo)
  end

  private

  def resolve_usuario(uid, email, usuario_doc = nil)
    return usuario_doc if usuario_doc

    doc = @firestore.buscar_usuario(uid, email)
    raise ArgumentError, "Usuario nao encontrado no banco." unless doc

    doc
  end

  def validar_submissao(payload)
    raise ArgumentError, "Payload de submissao nao enviado." if payload.blank?
    raise ArgumentError, "Titulo e tipo sao obrigatorios." if payload["titulo"].blank? || payload["tipo"].blank?
    raise ArgumentError, "Horas deve ser maior que zero." if payload["horas"].to_i <= 0
  end

  def para_atividade(doc)
    data = doc[:data]
    {
      id: FirestoreKeyNormalizer.fetch_field(data, "id_solicitacao") || doc[:id],
      titulo: texto(FirestoreKeyNormalizer.fetch_field(data, "titulo_atividade")),
      tipo: texto(FirestoreKeyNormalizer.fetch_field(data, "categoria")),
      categoria: texto(FirestoreKeyNormalizer.fetch_field(data, "categoria")),
      horas: FirestoreKeyNormalizer.fetch_numeric(data, "horas_informadas"),
      horasAprovadas: FirestoreKeyNormalizer.fetch_numeric(data, "horas_aprovadas"),
      status: normalizar_status(texto(FirestoreKeyNormalizer.fetch_field(data, "status"))),
      comprovanteUrl: texto(FirestoreKeyNormalizer.fetch_field(data, "url_certificado")),
      data: formatar_data(FirestoreKeyNormalizer.fetch_field(data, "data_envio")),
      justificativaCoordenador: texto(FirestoreKeyNormalizer.fetch_field(data, "justificativa_coordenador"))
    }
  end

  def primeiro_vinculo(usuario)
    data = usuario[:data]
    raw = FirestoreKeyNormalizer.fetch_field(data, "vinculo")

    if raw.is_a?(Hash)
      return stringify_hash(raw)
    end
    if raw.is_a?(Array) && raw.first.is_a?(Hash)
      return stringify_hash(raw.first)
    end

    vinculo = {
      "id_curso" => texto(FirestoreKeyNormalizer.fetch_field(data, "id_curso", "curso")),
      "matricula" => texto(FirestoreKeyNormalizer.fetch_field(data, "matricula")),
      "ch_total_exigida" => FirestoreKeyNormalizer.fetch_field(data, "ch_total_exigida")
    }

    if vinculo["id_curso"].present? || vinculo["matricula"].present?
      return vinculo
    end

    raise ArgumentError, "Usuario sem vinculo no curso."
  end

  def stringify_hash(hash)
    hash.each_with_object({}) { |(k, v), h| h[k.to_s] = v }
  end

  def resolver_categoria(categoria, tipo)
    return capitalizar_categoria(categoria) if categoria.present?

    t = tipo.to_s.downcase
    return "Pesquisa" if t.include?("pesquisa")
    return "Ensino" if t.include?("monitoria") || t.include?("curso")

    "Extensao"
  end

  def capitalizar_categoria(categoria)
    c = categoria.to_s.strip.downcase
    return "Ensino" if c.start_with?("ens")
    return "Pesquisa" if c.start_with?("pes")

    "Extensao"
  end

  def normalizar_status(status)
    return "INDEFERIDO" if status.to_s.upcase == "REPROVADO"

    status.presence&.upcase || "PENDENTE"
  end

  def formatar_data(data_envio)
    return data_envio.to_time.to_s if data_envio.respond_to?(:to_time)
    return data_envio.to_s if data_envio

    ""
  end

  def curso_do_vinculo(vinculo)
    id_curso = texto(FirestoreKeyNormalizer.fetch_field(vinculo, "id_curso", "curso"))
    return id_curso if id_curso.blank?

    curso = @firestore.buscar_curso(id_curso)
    if curso
      nome = texto(FirestoreKeyNormalizer.fetch_field(curso[:data], "nome_curso", "nome"))
      return nome if nome.present?
    end
    id_curso
  end

  def texto(value)
    value.nil? ? "" : value.to_s
  end
end

File: backend/src/services/alunoService.js
const { IllegalArgumentError } = require('../middleware/errorHandler');

class AlunoService {
  constructor(firestoreService, fileUploadService) {
    this.firestoreService = firestoreService;
    this.fileUploadService = fileUploadService;
    // PerfilService fica em serviços; aqui só precisa resolver perfil quando chamado por /api/usuario
    // Para manter sem acoplamento circular, a rota /api/usuario usa perfilResolver.
    this.perfilResolverFn = null;
  }

  // setado pelo getServices
  setPerfilResolver(fn) {
    this.perfilResolverFn = fn;
  }

  perfilResolver(uid, email) {
    return this.perfilResolverFn(uid, email);
  }

  async buscarPerfil(uid, email) {
    const usuario = await this.buscarUsuario(uid, email);
    const primeiroVinculo = this.primeiroVinculo(usuario);

    return {
      uid: usuario.get('uid'),
      nome: usuario.get('nome'),
      email: usuario.get('email'),
      telefone: '',
      ingresso: '',
      curso: await this.cursoDoVinculo(primeiroVinculo),
      departamento: '',
      matricula: this.texto(primeiroVinculo.matricula)
    };

  }


  async buscarResumo(uid, email) {
    const usuario = await this.buscarUsuario(uid, email);
    const vinculo = this.primeiroVinculo(usuario);
    const atividades = await this.listarHistorico(uid, email);

    const aprovadas = atividades.filter((a) => a.status === 'APROVADO').length;
    const pendentes = atividades.filter((a) => a.status === 'PENDENTE').length;
    const indeferidas = atividades.filter((a) => a.status === 'REPROVADO').length;

    const saldos = this.castMap(vinculo.saldos);
    const horasConcluidas =
      this.asInt(saldos.ensino) + this.asInt(saldos.pesquisa) + this.asInt(saldos.extensao);

    let horasNecessarias = this.asInt(vinculo.ch_total_exigida);
    if (horasNecessarias <= 0) horasNecessarias = 200;

    const percentualConcluido = Math.min(100, Math.floor((horasConcluidas * 100) / horasNecessarias));

    return {
      curso: await this.cursoDoVinculo(vinculo),
      horasConcluidas,
      horasNecessarias,
      percentualConcluido,
      aprovadas,
      pendentes,
      indeferidas,
      totalAtividades: atividades.length
    };
  }

  async listarHistorico(uid, email) {
    const usuario = await this.buscarUsuario(uid, email);
    const uidAluno = usuario.id;

    const docs = await this.firestoreService.listarSolicitacoesPorAluno(uidAluno);
    const atividades = docs.map((doc) => this.paraAtividade(doc));

    const all = await Promise.all(atividades);
    all.sort((a, b) => (b.data || '').localeCompare(a.data || ''));
    return all;
  }

  async listarRecentes(limite, uid, email) {
    const all = await this.listarHistorico(uid, email);
    return all.slice(0, limite);
  }

  async submeterAtividade(payload, uid, email) {
    this.validarSubmissao(payload);
    const usuario = await this.buscarUsuario(uid, email);
    const vinculo = this.primeiroVinculo(usuario);

    const idCursoPayload = this.cursoId(payload.idCurso);
    const idCurso = this.isBlank(idCursoPayload) ? this.cursoId(vinculo.id_curso || vinculo.curso) : idCursoPayload;
    const categoria = this.resolverCategoria(payload.categoria, payload.tipo);

    const comprovanteUrl = await this.fileUploadService.uploadDataUrl(payload.comprovanteUrl, usuario.id);

    const doc = {
      uid_aluno: usuario.id,
      id_curso: idCurso,
      titulo_atividade: String(payload.titulo).trim(),
      categoria,
      horas_informadas: payload.horas,
      horas_aprovadas: 0,
      status: 'PENDENTE',
      url_certificado: comprovanteUrl,
      data_evento: payload.data ? String(payload.data).trim() : '',
      data_envio: new Date().toISOString(),
      justificativa_coordenador: ''
    };

    const salvo = await this.firestoreService.salvarSolicitacao(doc);
    return this.paraAtividade(salvo);
  }

  validarSubmissao(payload) {
    if (!payload) throw new IllegalArgumentError('Payload de submissao nao enviado.');
    if (this.isBlank(payload.titulo) || this.isBlank(payload.tipo)) {
      throw new IllegalArgumentError('Titulo e tipo sao obrigatorios.');
    }
    if (payload.horas <= 0) {
      throw new IllegalArgumentError('Horas deve ser maior que zero.');
    }
  }

  async buscarUsuario(uid, email) {
    const porUid = await this.firestoreService.buscarUsuarioPorUid(uid);
    if (porUid) return porUid;

    const byEmail = await this.firestoreService.buscarUsuarioPorEmail(email);
    if (!byEmail) throw new IllegalArgumentError('Usuario nao encontrado no banco.');
    return byEmail;
  }

  paraAtividade(doc) {
    const categoriaRaw = doc.get('categoria');
    const statusRaw = doc.get('status');

    return {
      id: doc.get('id_solicitacao') != null ? doc.get('id_solicitacao') : doc.id,
      titulo: this.texto(doc.get('titulo_atividade')),
      tipo: this.texto(categoriaRaw),
      categoria: this.texto(categoriaRaw),
      horas: this.asInt(doc.get('horas_informadas')),
      horasAprovadas: this.asInt(doc.get('horas_aprovadas')),
      status: this.normalizarStatus(this.texto(statusRaw)),
      comprovanteUrl: this.texto(doc.get('url_certificado')),
      data: this.formatarData(doc.get('data_evento') || doc.get('data_envio')),
      justificativaCoordenador: this.texto(doc.get('justificativa_coordenador'))
    };
  }

  primeiroVinculo(usuarioDoc) {
    const vinculoRaw = usuarioDoc.get('vinculo');

    if (vinculoRaw && typeof vinculoRaw === 'object' && !Array.isArray(vinculoRaw)) {
      return { ...vinculoRaw };
    }
    if (Array.isArray(vinculoRaw) && vinculoRaw.length > 0 && typeof vinculoRaw[0] === 'object') {
      return { ...vinculoRaw[0] };
    }

    throw new IllegalArgumentError('Usuario sem vinculo no curso.');
  }

  castMap(value) {
    if (value && typeof value === 'object' && !Array.isArray(value)) return { ...value };
    return {};
  }

  asInt(value) {
    if (typeof value === 'number') return Math.trunc(value);
    if (value == null) return 0;
    const n = Number(value);
    return Number.isFinite(n) ? Math.trunc(n) : 0;
  }

  texto(value) {
    return value == null ? '' : String(value);
  }

  resolverCategoria(categoria, tipo) {
    if (!this.isBlank(categoria)) return this.capitalizarCategoria(categoria);
    const t = tipo == null ? '' : String(tipo).toLowerCase();
    if (t.includes('pesquisa')) return 'Pesquisa';
    if (t.includes('monitoria') || t.includes('curso')) return 'Ensino';
    return 'Extensao';
  }

  capitalizarCategoria(categoria) {
    const c = String(categoria).trim().toLowerCase();
    if (c.startsWith('ens')) return 'Ensino';
    if (c.startsWith('pes')) return 'Pesquisa';
    return 'Extensao';
  }

  normalizarStatus(status) {
    if ('REPROVADO' === String(status)) return 'INDEFERIDO';
    return status == null || status === '' ? 'PENDENTE' : String(status).toUpperCase();
  }

  formatarData(dataEnvio) {
    return dataEnvio ? String(dataEnvio) : '';
  }

  async cursoDoVinculo(vinculo) {
    const cursoRaw = vinculo.id_curso || vinculo.curso;
    const nomeNoVinculo = this.cursoNome(cursoRaw);
    if (!this.isBlank(nomeNoVinculo)) return nomeNoVinculo;

    const idCurso = this.cursoId(cursoRaw);
    const c = await this.firestoreService.buscarCurso(idCurso);
    if (c) {
      const nome = this.texto(c.get('nome_curso'));
      if (!this.isBlank(nome)) return nome;
    }
    return idCurso;
  }

  cursoId(value) {
    if (value == null) return '';
    if (typeof value === 'object') {
      if (Array.isArray(value)) return value.length ? this.cursoId(value[0]) : '';
      return this.texto(value.id_curso || value.idCurso || value.id || value.codigo || value.codigoCurso).trim();
    }
    return this.texto(value).trim();
  }

  cursoNome(value) {
    if (value == null || typeof value !== 'object') return '';
    if (Array.isArray(value)) return value.length ? this.cursoNome(value[0]) : '';
    return this.texto(value.nome_curso || value.nomeCurso || value.nome || value.descricao);
  }

  isBlank(valor) {
    return valor == null || String(valor).trim().length === 0;
  }
}

module.exports = AlunoService;

